From 85fded65b3684618bae0219c94b1e939dc09d096 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Tue, 7 May 2019 12:10:46 +0200
Subject: [PATCH 191/345] MIPS: Fix demand activation of OCTEON CVMSEG region.

OCTEON III processors can use several addresses in the CVMSEG_IO
range, so enable them all.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Carlos Munoz <cmunoz@caviumnetworks.com>
---
 arch/mips/kernel/unaligned.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/arch/mips/kernel/unaligned.c b/arch/mips/kernel/unaligned.c
index 9b32207b6eda..c64d40e9ecbf 100644
--- a/arch/mips/kernel/unaligned.c
+++ b/arch/mips/kernel/unaligned.c
@@ -2326,7 +2326,8 @@ asmlinkage void do_ade(struct pt_regs *regs)
 	mm_segment_t seg;
 #if defined(CONFIG_CPU_CAVIUM_OCTEON) && (CONFIG_CAVIUM_OCTEON_CVMSEG_SIZE > 0)
 	const unsigned long CVMSEG_BASE	= 0xffffffffffff8000ul;
-	const unsigned long CVMSEG_IO	= 0xffffffffffffa200ul;
+	const unsigned long CVMSEG_IO		= 0xffffffffffffa000ul;
+	const unsigned long CVMSEG_IO_END	= 0xffffffffffffc000ul;
 	u64 cvmmemctl			= __read_64bit_c0_register($11, 7);
 	unsigned long cvmseg_size	= (cvmmemctl & 0x3f) * 128;
 #endif
@@ -2343,9 +2344,8 @@ asmlinkage void do_ade(struct pt_regs *regs)
 	 * addresses, we enable the region and size it to match the
 	 * app.
 	 */
-	if ((regs->cp0_badvaddr == CVMSEG_IO) ||
-	    ((regs->cp0_badvaddr >= CVMSEG_BASE) &&
-	     (regs->cp0_badvaddr < CVMSEG_BASE + cvmseg_size))) {
+	if ((regs->cp0_badvaddr >= CVMSEG_IO && regs->cp0_badvaddr < CVMSEG_IO_END) ||
+	    (regs->cp0_badvaddr >= CVMSEG_BASE && regs->cp0_badvaddr < CVMSEG_BASE + cvmseg_size)) {
 		preempt_disable();
 		cvmmemctl = __read_64bit_c0_register($11, 7);
 		/* Make sure all async operations are done */
-- 
2.25.1

